Pass JSX props to function components instead of directly assigning them to the element

Kir-Antipov opened the pull request:
## Description:

`dom-chef` didn't pass props to the element's constructor.

## MWE:

```jsx
function Element({ value }) {
    return <div>{value}</div>;
}
document.body.append(<Element value='Hello world'/>);
```

## Old behavior:
```js
// Uncaught TypeError: (destructured parameter) is undefined
```

## Expected/New behavior:
```html
<div>Hello world</div>
```

## Breaking changes

```jsx
const Element = () => <div/>;

document.body.append(<Element className='foo'/>);
```

Previously, this would produce the following element:

```html
<div class='foo'></div>
```

But that's not how things should be. The correct (and current) output is:

```html
<div></div>
```

Kir-Antipov commented:
P.S. - `ESM / Webpack (pull_request)` failure is not my fault :)

fregante commented:
> `dom-chef` didn't pass props to the element's constructor.

This was a choice that was made when adding support for functions: https://github.com/vadimdemedes/dom-chef/pull/59

> My resolution is that if one wants `Component` to receive `{prop: 1}` they should not use JSX.


Therefore, what's preventing you from just writing this?

```diff
- document.body.append(<Element value='Hello world'/>);
+ document.body.append(Element({value: 'Hello world'});
```

Kir-Antipov commented:
> This was a choice that was made when adding support for functions

@fregante, that's interesting

> In dom-chef's world this wouldn't make sense

And I really cannot understand why this makes sense in React, but does not in "dom-chef's" world. JSX is JSX ¯\\_(ツ)_/¯

> Therefore, what's preventing you from just writing this?

And what's preventing me from writing it that way I did? :)

---------

Well, that's my point:

As you said,

> But probably dom-chef should not become its own framework.

It's truly good thought. So, what do users expect when using dom-chef? They expect that dom-chef will be similar to React they already know (except for the main thing, which is described in the library description: we get the good old HTMLElement as the output). And so did I, especially considering the fact that the Readme says

> To improve **compatibility with React** components, [...]

So React-compatibility is still in mind, but one of the most important concepts is somehow missing.

Returning to the question why I can't just use a function instead of JSX, I want to add: it's inconvenient and leads to a lot of errors. Imagine that you are writing a large component that consists of many standard elements as well as other components. You will constantly have to keep in mind that for some strange reason you cannot use JSX for components, you need to call the function manually. And you will surely forget it out of habit, so you won't get what you expected, and then you'll just sigh sadly and go to correct what you have written for the twentieth time. And that's not to mention that the code gets dirtier and less readable:

```jsx
const result = Foo({ bar: baz });
const innerFoo = Foo({ baz: bar });
innerFoo.append(
    <span>Hello</span>
);
result.append(
    <div className='foo'>Bar</div>,
    innerFoo,
    (
        <form>
            {Foo({ bar: bar })}
            <button type='submit'>Save</button>
        </form>
    )
);
return result;
```

instead of

```jsx
return (
    <Foo bar={baz}>
        <div className='foo'>Bar</div>
        <Foo baz={bar}>
            <span>Hello</span>
        </Foo>
        <form>
            <Foo bar={bar}/>
            <button type='submit'>Save</button>
        </form>
    </Foo>
);
```

I hope it's clear now

fregante commented:
> And I really cannot understand why this makes sense in React, but does not in "dom-chef's world". JSX is JSX ¯\_(ツ)_/¯

I assume React components natively handle props and children, while the "Element-returning functions" as described by dom-chef don't.

What would `Foo` have to do if I wanted the equivalent of the current `<Foo className="baz" data-name="bar">`?

Kir-Antipov commented:
> I assume React components natively handle props and children, while the "Element-returning functions" as described by dom-chef don't.

Only standard elements are processed natively. Once we define our own component, we decide what to do with the props and/or children\* (even if we delegate some of these responsibilities to the chosen framework, this is still our choice.).

*P.S. - You just reminded me to add a list of child elements to the props.

----------------

> What would `Foo` have to do if I wanted the equivalent of the current `<Foo className="baz" data-name="bar">`?

As I said above, we ourselves decide what to do with the props, so if we want such an output:

```html
<div class='baz' data-name='bar'>
    <span>Hello world</span>
</div>
```

we describe it something like this:

```jsx
const Foo = (props) => {
    return <div className={props.className} data-name={props.name}>{props.children}</div>;
}

const element = (
    <Foo className='baz' name='bar'>
        <span>Hello world</span>
    </Foo>
);
```

fregante commented:
This makes sense. In Refined GitHub we're [using a React module](https://github.com/sindresorhus/refined-github/blob/600ecbfd830702181cd800516630680a8a0507fb/source/github-widgets/notice-bar.tsx#L3) by replacing react with dom-chef at build time. It worked until now but custom props weren't handled (which was ok in this case).

I think 2 changes are necessary now:

- [ ] I think dom-chef should handle or ignore the `children` prop when passed as an attribute
- [ ] The docs should be updated to reflect this change

However there's a big convenience that now will be lost:

`<Icon className="mr-1">` would _add_ the `mr-1` class to whatever Icon specifies. After this PR the default classes will be overridden.

Kir-Antipov commented:
> I think dom-chef should handle or ignore the `children` prop when passed as an attribute

@fregante, it's already done: `children` prop **always** contains a list of the component's child elements (or an empty array)

> The docs should be updated to reflect this change

I agree. I'll try to reflect the changes in the Readme :) 

> However there's a big convenience that now will be lost

*Well yes, but actually no*. There's the reason why it's not a thing in React: React, dom-chef or any other framework/library knows nothing about user-defined components, so it is completely unclear where the specified class should be added and whether it should be added at all (because components are usually self-contained and very rarely have the functionality that allows to pass additional classes to them).

People who for some reason relied on this feature may simply not update to the newer version. But the best solution would be to bring their code base in line with generally accepted standards :)

fregante commented:
> it's already done: `children` prop **always** contains a list of the component's child elements (or an empty array)

No it doesn't, at least in the code that JSX produces:

<img width="685" alt="" src="https://user-images.githubusercontent.com/1402241/104829166-abe01600-5836-11eb-8df9-0cffdec10356.png">

Props here is `{className: "class"}`, not `{className: "class", children: ["Title"]}`

`dom-chef` _does not_ handle the `children` property because it never generate it (since, again, JSX doesn't do that):

https://github.com/vadimdemedes/dom-chef/blob/fc72a4e86ec3ca9c037635c70366a1e4dff21847/index.ts#L119-L134

Kir-Antipov commented:
@fregante, You overtook me by a minute, cause I was about to push the test xD

The correct test looks like this:
```jsx
test('`children` prop always contains a list of child nodes and ignores `children` attribute', t => {
	const Icon = ({children}: {children: Node | string}) => <i>{children}</i>;

	const elementWithChildren = <Icon><span>Hello world</span></Icon>;
	const elementWithoutChildren = <Icon children="Hello world"/>;

	t.is(elementWithChildren.outerHTML, '<i><span>Hello world</span></i>');
	t.is(elementWithoutChildren.outerHTML, '<i></i>');
});
```

fregante commented:
> The correct test looks like this:

Your test shows what happens when one manually sets a `children` string attribute, which they don't. We need to handle the `children` prop that React and now dom-chef generate. This prop is an array of DOM elements and strings, not a plain string.

Kir-Antipov commented:
> You can see this in the test I just added, it generate this line:

@fregante, of course it does, cause your JSX looks like:

```jsx
<i title="icon" {...props}/>
```

Does `props` have `children` property? Yep. Why did you expect this output?

```html
<i title="icon" class="yellow">Submarine</i>
```

Kir-Antipov commented:
> This prop is an array of DOM elements and strings, not a plain string.

Yeah, I know, and it is. But Typescript didn't believe me :)

Kir-Antipov commented:
If you want such an output:

```jsx
<i title="icon" class="yellow">Submarine</i>
```

Your function should look like this:

```jsx
const Icon = (props: Record<string, any>) => (
    <i title="icon" className={props.className}>
        {props.children}
    </i>
);

const element = <Icon className="yellow">Submarine</Icon>;
```

fregante commented:
> ```js-jsx
> ignores `children` attribute
> ```

This is also incorrect. The `children` attribute is not ignored, it's set onto intrinsic elements but it's override if the component already one: https://codepen.io/fregante/pen/bGwOZVq

<img width="232" alt="" src="https://user-images.githubusercontent.com/1402241/104829416-2d38a800-5839-11eb-8457-eddab8c09405.png">

Kir-Antipov commented:
Oh, ok
I've got the point

fregante commented:
> Your function should look like this:

That's just one way to do it. If you want to transfer all props, you can use the rest operator: https://reactjs.org/docs/jsx-in-depth.html#spread-attributes

fregante commented:
@yakov116 @notlmn heads up for this upcoming change

Kir-Antipov commented:
@fregante, done. Also I've added test based on [your example](https://codepen.io/fregante/pen/bGwOZVq).

Now I'm going to update the Readme

Kir-Antipov commented:
I couldn't resist and added support for components that look just like React ones:

```jsx
import React from 'dom-chef';

class Icon extends React.Component {
	render() {
		return <i {...this.props}></i>;
	}
}
```

This is done for better compatibility with React, as well as for component inheritance:

```jsx
import React from 'dom-chef';

class Button extends React.Component {
  onClick() {
    console.log('Click');
  }

	render() {
		return <button {...this.props} onClick={this.onClick}/>;
	}
}

class CoolerButton extends Button {
  render() {
    return (
      <cool-wrapper>
        {super.render()}
      </cool-wrapper>
    );
  }
}
```

fregante commented:
> I couldn't resist and added support for components that look just like React ones:

I don't think that's gonna make the cut. Perhaps we can discuss it in a new PR

Kir-Antipov commented:
@fregante, why not discuss it here?)

This functionality doesn't change the existing behavior, it just expands the use cases.

If we restrict users solely to functions that return elements, users lose the ability to inherit from existing components, which leads to a lot of inconvenience.
Want to reuse an existing layout, but change the handlers? Sorry, pal, deal with this yourself
Don't want to pile up the handler functions outside the component?  Sorry, but classes aren't available for you, welcome to 2005
And so on

So it's not about React-compatibility only (but porting components from React will become easier too), first of all, it's about the usability of dom-chef

Few lines of code here, but so much convenience for everyone

fregante commented:
Please send it as a separate PR after this, it will be a non-breaking change anyway.

Kir-Antipov commented:
On the other hand, this PR was about props-handling, so, yeah, agree. Should I include updated readme as part of the next PR, or should I update it for this one at first, and then for the next?

fregante commented:
> should I update it for this one at first

fregante commented:
Check out https://github.com/Kir-Antipov/dom-chef/pull/1, the tests pass and it doesn't look like that code is needed

fregante commented:
Note to self: https://github.com/vadimdemedes/dom-chef/pull/62 needs to be merged before the next version as well

fregante commented:
If you could avoid force-pushing in the future it'd be easier to follow the changes. Commits will be squashed anyway so commits don't end up on master

Kir-Antipov commented:
> If you could avoid force-pushing in the future it'd be easier to follow the changes. Commits will be squashed anyway so commits don't end up on master

Ooh ok tried to keep your commit history cleaner

fregante commented:
Let me know if you see any other issues here

fregante commented on readme.md:
````diff
@@ -152,38 +152,50 @@ If element names start with an uppercase letter, `dom-chef` will consider them a
 ```
 
-To improve compatibility with React components, `dom-chef` will pass the function's `defaultProps` property to itself (if present). Note that specifying attributes won't override those defaults, but instead set them on the resulting element:
+To improve compatibility with React components, `dom-chef` will pass the function's `defaultProps` property to itself (if present). Note that specifying attributes will override those defaults:
````
Given that this now follows React’s behavior it doesn't need to be explained:

```suggestion
[`defaultProps`](https://medium.com/@pojotorshemi/react-basis-default-props-and-typechecking-with-proptypes-4ddf02d15992) is also supported and will be included as props to the function as you'd expect:
```

However I can't find an official documentation for it. Do you know a better resource?

fregante commented on tests/index.tsx:
```diff
@@ -473,17 +493,39 @@ test('element created by function with existing children and attributes', t => {
 
 test('element created by function with combined children and attributes', t => {
-	const Icon = () => <i className="sweet">Gummy <span>bears</span></i>;
+	const Icon = ({children}: {children: Node[]}) => <i className="sweet">Gummy <span>bears</span>{children}</i>;
```
Not required, but can you find the correct type here? I think it would be `Attributes`, `Props` or something like it. 
```suggestion
	const Icon = ({children}: Attributes) => <i className="sweet">Gummy <span>bears</span>{children}</i>;
```

Same goes for `Record<string, any>` used below

Kir-Antipov replied:
I don't think it really matters, since React Component defined kinda like this:

```jsx
class Component<P = {}> {
    props: P;
    constructor(props: P) {
        this.props = props;
    }
}
```

So `props` can be of any type by definition. `Record<string, any>` suits it perfectly, imho

fregante replied:
That's a class; function components seem to use `React.FC`: https://www.carlrippon.com/Different-ways-to-strongly-type-function-component-props-with-typescript/

> `FC` includes the children prop type, so you don’t have to define it explicitly

fregante replied:
Speaking of which, it appears that the types might be broken for function components at the moment: https://github.com/sindresorhus/refined-github/pull/3904#discussion_r560722216

Could you try v5.0.0-2 to verify if it's just Refined GitHub or if it's `dom-chef`

fregante replied:
Possible example found here: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/41419

fregante reviewed (approved):
I think that other than the mentioned test and this documentation update it should be good to go. Let me know if you see any issues

fregante commented:
This has been prereleased as v5.0.0-2 to allow testing

Kir-Antipov commented:
> Regarding this change: can you add a test to ensure this? Currently it passes regardless of this commit.

Yup, you accidentally swapped the blocks while inlining the functions, so props-forming logic was slightly broken. The test I added earlier passed as it relied on the **h1** elements xD

Kir-Antipov commented:
May I also ask you not to change this line?

```json
"pretest:ava": "rollup -p json -p \"typescript={declaration: false, outDir: './tests'}\" --format cjs --dir tests tests/index.tsx",
```

The single quoted version works fine for me on macOS (and obviously on Linux, since the checks are run on Ubuntu), but doesn't work at all on Windows. Double quotes work for every OS, so it is worth keeping them for the convenience of future contributors

fregante commented on index.ts:
```diff
@@ -8,9 +8,11 @@ svgTags.delete('iframe');
+	children?: Node[];
+};
+
+type FunctionComponent = ((props?: any) => Element | DocumentFragment) & {
```
Since we already install `@types/react`, maybe we can just use `React.FC` instead of this type.
Maybe the `declare global` block below fixes the return type of React.FC to be a DOM element as well.

Kir-Antipov replied:
Nope, doesn't work

fregante commented:
> May I also ask you not to change this line?

Yeah that was lost in the merge since I changed the build. As long as the configuration object works with single quotes that's fine.

fregante commented:
> Regarding this change: can you add a test to ensure this? Currently it passes regardless of this commit.

I checked your last test and it correctly fails on the previous version. 👍

Kir-Antipov commented:
@fregante, how's it going? :) Do we need something else?

fregante commented:
FunctionComponent are still too generic. Children and className should always exist. Even if React.FC can’t be used directly we need to provide an equivalent type, at least as a guideline what `props` should be so that we don’t have to give weird inline types like we do currently in the tests. 

Also this PR is already available on npm so it can already be used without having to rush its official release.

fregante commented:
Refer to the open review threads, one of them points to https://github.com/sindresorhus/refined-github/pull/3904#discussion_r560722216, which shows a real React function component being of type `any` for some reason, because React.FC is broken. If this PR is specifically to improve React compatibility I think it’s worth investigating why the types for them are `any`

Kir-Antipov commented:
> and className should always exist

Nope, it should not

> Even if React.FC can’t be used directly

It can. Well, I removed `FunctionComponent` and replaced it with `React.FC`. Along the way, the problem you are talking about was resolved (the type of elements created using `React.FC` is `JSX.Element` as expected).

Kir-Antipov commented:
```jsx
function Icon({ className }: { className: string }) {
    const icon = document.createElement('i');
    icon.classList.add(className);
    return icon;
}

...

const icon = <Icon className="icon"/>; // JSX.Element
```

Now TypeScript has no questions about such code too, it transpiles successfully and the type of the resulting element is recognized as `JSX.Element`.

Any of these examples are valid now:
```jsx
function Icon({ className }: { className: string }) {
    const icon = document.createElement('i');
    icon.classList.add(className);
    return icon;
}

// ...

const Icon = ({ className }: { className: string }) => { 
    const icon = document.createElement('i');
    icon.classList.add(className);
    return icon;
};

// ...

const Icon: React.FC<{ className: string }> = ({ className }) => {
    const icon = document.createElement('i');
    icon.classList.add(className);
    return icon;
};

// And so on
```

fregante commented:
Why close it? I'd still like this feature, I'm just not convinced about the types. I was hoping for someone to comment

fregante commented:
This article may be of help here: https://fettblog.eu/typescript-react-why-i-dont-use-react-fc/

So maybe `React.FC` shouldn't be used after all? I don't have time to look into this right now, but hopefully within the end of the year? heh heh.

Kir-Antipov commented:
> Why close it?

@fregante, welp, I just have no patience to sit here for like another year or two :)
I was already ready to start writing my own solution, but in the process of choosing a name for it, I discovered [`jsx-dom`](https://github.com/proteriax/jsx-dom). It has a cleaner codebase *(so it's much-much easier to make changes without merge conflicts, e.g. 3 of my PRs coexisted there at the same time without any problems, while the all-in-one-file `dom-chef` doesn't allow such a thing)* and a lot of cool features. So I've just fixed few small problems and added support for class components and SVG presentation attributes, and now it's perfect for my purposes. So I'll just stick with `jsx-dom` for my current and future projects

> I'd still like this feature

If you really want this feature, I can reopen the PR and recheck "Allow edits by maintainers", but it's all up to you now, since I found what I was looking for, sorry

fregante commented:
Alright, I’ll close this for now

Kir-Antipov commented:
@fregante, since you've banned me and I can't answer in that issue (proteriax/jsx-dom#47) directly, I'll make it here

Judging by an extremely unpleasant tone of your last message, I managed to hurt you somehow. Since I usually try to be friendly with everyone, I would like to apologize for whatever I said/made. It was **not** intended to offend you in any way. I guess, this:

> Why do you use TypeScript if you check your code's vitality at runtime? Maybe you should switch to plain JavaScript?

was a problem. It's **impersonal** sentence, I didn't try to say something like: "You're bad TypeScript programmer" - no, my message had completely different meaning, more like: "I don't feel it's TypeScript way to do things" I tried to point out that if we're using TypeScript, we **need** this problems to ensure that everything's ok before transpilation, and if we're using JS than we do not need them and, well, there've already disappeared. 

So, once more, I'm really sorry if you took it too much to your heart

But I can't help but say that banning a person for one phrase, which doesn't contain any intentional insults, because of overthinking, is really childish. And if this statement offends you, I'm not sorry for it, since it's my reasonable point of view. I thought of you more as a grown man, when I first encountered dom-chef

Welp, try to see the best in people, don't overthink and relax more often. Peace

fregante commented:
Strange to see this after your relaxing message

<img width="272" alt="Screen Shot 8" src="https://user-images.githubusercontent.com/1402241/111016206-1febf580-8372-11eb-98d8-0904cf6d3c20.png">
